File: CoreJava/RicksGuitar-app/src/com/techlabs/ricksguitar/guitars/Inventory.java
package com.techlabs.ricksguitar.guitars;

import java.util.*;

public class Inventory {
	private List<Guitar> guitars = new LinkedList<Guitar>();
	private Iterator<Guitar> iterator = guitars.iterator();

	public void addGuitar(String serialNumber, double price, GuitarSpec spec) {
		guitars.add(new Guitar(serialNumber, price, spec));
	}

	public Guitar getGuitar(String serialNumber) {
		while (iterator.hasNext()) {
			if (iterator.next().getSerialNumber().equals(serialNumber))
				return iterator.next();
		}
		return null;
	}

	public List<Guitar> search(GuitarSpec searchSpec) {
		List<Guitar> matchingGuitars = new LinkedList<Guitar>();

		while (iterator.hasNext()) {
			if (iterator.next().getSpec().matches(searchSpec))
				matchingGuitars.add(iterator.next());
		}

		return matchingGuitars;
	}
}

File: DotNET/C#/RicksGuitarApp/RicksGuitarApp/Inventory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace RicksGuitarApp
{
    class Inventory
    {
        private static List<Guitar> guitars = new List<Guitar>();
        private IEnumerator<Guitar> iterator = guitars.GetEnumerator();

        public void AddGuitar(String serialNumber, double price, GuitarSpec spec)
        {
            guitars.Add(new Guitar(serialNumber, price, spec));
        }

        public Guitar GetGuitar(String serialNumber)
        {
            using (iterator)
            {
                while (iterator.MoveNext())
                {
                    if (iterator.Current.SerialNumber.Equals(serialNumber))
                        return iterator.Current;
                }
            }
            return null;
        }

        public List<Guitar> search(GuitarSpec searchSpec)
        {
            List<Guitar> matchingGuitars = new List<Guitar>();

            foreach (Guitar guitar in guitars)
            {
                if (guitar.Spec.matches(searchSpec))
                    matchingGuitars.Add(iterator.Current);
            }
            return matchingGuitars;
        }
    }
}
